packages/block-library/src/page-list/convert-to-links-modal.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * WordPress dependencies
+ */
+import { Button, Modal } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { useState } from '@wordpress/element';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { store as coreDataStore } from '@wordpress/core-data';
+import { createBlock as create } from '@wordpress/blocks';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+
+const PAGE_FIELDS = [ 'id', 'title', 'link', 'type', 'parent' ];
+const MAX_PAGE_COUNT = 100;
+
+export const convertSelectedBlockToNavigationLinks = ( {
+	pages,
+	clientId,
+	replaceBlock,
+	createBlock,
+	setIsConverting,
+} ) => () => {
+	if ( ! pages ) {
+		return;
+	}
+
+	const linkMap = {};
+	const navigationLinks = [];
+	setIsConverting( true );
+	pages.forEach( ( { id, title, link: url, type, parent } ) => {
+		// See if a placeholder exists. This is created if children appear before parents in list
+		const innerBlocks = linkMap[ id ]?.innerBlocks ?? [];
+		linkMap[ id ] = createBlock(
+			'core/navigation-link',
+			{
+				id,
+				label: title.rendered, //TODO: raw or rendered?
+				url,
+				type,
+				kind: 'post-type',
+			},
+			innerBlocks
+		);
+
+		if ( ! parent ) {
+			navigationLinks.push( linkMap[ id ] );
+		} else {
+			if ( ! linkMap[ parent ] ) {
+				// Use a placeholder if the child appears before parent in list
+				linkMap[ parent ] = { innerBlocks: [] };
+			}
+			const parentLinkInnerBlocks = linkMap[ parent ].innerBlocks;
+			parentLinkInnerBlocks.push( linkMap[ id ] );
+		}
+	} );
+
+	replaceBlock( clientId, navigationLinks );
+	setIsConverting( false );
+};
+
+export default function ConvertToLinksModal( { onClose, clientId } ) {
+	const [ isConverting, setIsConverting ] = useState( false );
+	const { pages, pagesFinished } = useSelect(
+		( select ) => {
+			const { getEntityRecords, hasFinishedResolution } = select(
+				coreDataStore
+			);
+			const query = [
+				'postType',
+				'page',
+				{
+					per_page: MAX_PAGE_COUNT,
+					_fields: PAGE_FIELDS,
+					orderby: 'menu_order',
```

Comment:
Funny that this seems to return the pages in more or less oldest to most recent order, whereas [the same parameter passed to `get_pages`](https://github.com/WordPress/gutenberg/blob/add/edit-page-list/packages/block-library/src/page-list/index.php#L152) returns a completely different order 😕

Reply:
This is because get_pages has a default order of `asc` and WP_Query (used by the REST API) has a default order of `desc`, I'll double check which one we need to respect for menu_order.

It looks like we can't expect a stable sort from using menu_order, when items are equal to each other, so I'll need another dimension to sort on. Do folks prefer title alphabetizing or something like post date?

Reply:
For what it's worth, I don't think it's crucial to get this 100% right in this one PR. It's can be a nice enhancement in the future, but the value being able to convert the page list outweighs the downside that sorting might not be perfect, especially paired with the fact that you get immediate access to mover controls right after the conversion.

Reply:
Looks like it won't be possible to provide a consistent ordering until we can implement https://core.trac.wordpress.org/ticket/39037. 

> Fwiw, I implemented it that way initially because performance in the editor with >100 or so pages was pretty horrific

Good to know @tellthemachines! I'll look into adding multiple orderby support in the REST API next, then start on some general performance profiling for Navigation Links.

I've incorporated feedback and added a few explanations in e35f132